### ISSUE TITLE
Test different variants of K8s Services in conformance testing

### ISSUE THREAD

**Issue by dprotaso** [kubernetes-sigs/gateway-api#1718]
**What would you like to be added**:

We should add conformance tests that ensure Gateway API works well with different variants of K8s Services, Endpoint, EndpointSlices

Short List Variants
- Service - with selector
- Service - with selector, [endpoint slice mirror disabled](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-mirroring)
- Headless Service - with manual endpoints (and their different types)
- Headless Service - with manual endpoint slices (and their different types - ie. FQDN)

**Why this is needed**:

Different platforms that build on-top of K8s will utilize Services in different ways to suit their needs - it would be great to know Gateway conformance  supports these variants. ie. some implementations currently don't track EndpointSlices which is pretty fundamental as they've been stable for numerous years

**Comment by dprotaso** [comment]
Related - https://github.com/kubernetes-sigs/gateway-api/pull/1816

@shaneutt Given `ExternalName` support is marked as `SHOULD NOT` - does it still warrant having an opt-in test? The purpose being for implementations to do support that type they'll have a way to verify it

**Comment by shaneutt** [comment]
> Related - #1816
> 
> @shaneutt Given `ExternalName` support is marked as `SHOULD NOT` - does it still warrant having an opt-in test? The purpose being for implementations (that support this type) having a common way to verify their support.

Because we consider it `Implementation-specific` it sorta suggests the _opposite_ of conformance: implementations can choose to implement that functionality in a way they deem appropriate and safe. I'm not saying that we should _never_ have conformance tests that cover `Implementation-specific` behaviors, but I'm hesitant to say we need one here and wonder if that would make it feel more like "extended". What specifically did you have in mind?

**Comment by dprotaso** [comment]
@Xunzhuo are you still working on this?

**Comment by dprotaso** [comment]
I've created a first pass of a test here - https://github.com/kubernetes-sigs/gateway-api/pull/2828

Currently the test exercises the following variants:

1. Service - Headless (spec.clusterIP=None) with Selector
1. Service - Headless no selector but with a manual Endpoint (addressType: IPv4)
1. Service - Headless no selector but with a manual EndpointSlice (addressType: IPv4)
1. Service - with no selector but with a manual Endpoint (addressType: IPv4)
1. Service - with no selector but with a manual EndpointSlice (addressType: IPv4)

**Comment by dprotaso** [comment]
Circling back what's done is

1. Service with Manual EndpointSlices 
2. Headless K8s Services with Manual Endpointslices
3. Headless Service

We dropped `Endpoint` testing (see: https://github.com/kubernetes-sigs/gateway-api/pull/2828#discussion_r1518382335). New features will be added to EndpointSlice. Also Endpoints get mirrored to EndpointSlices.

The test works with both IP6 and IP4 clusters - it had to be conditional on the environment.

**Comment by dprotaso** [comment]
What's sorta left to discuss is what do we do with:

1. Service - with selector, [endpoint slice mirror disabled](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-mirroring)
2. EndpointSlice with FQDN (using CoreDNS this becomes a CNAME - but it's sorta undefined) see: https://github.com/kubernetes/kubernetes/issues/114210

**Comment by dprotaso** [comment]
re: 1 - @robscott's comment from the PR was

> No, there's no reason for controllers to watch both Endpoints and EndpointSlice APIs. Gateway API is newer than the EndpointSlice API so hopefully all implementations are reading from EndpointSlices, but they definitely shouldn't be reading from both APIs.

Confirmed with @arkodg Envoy Gateway only watches endpoint slices 
Confirmed with @sunjayBhatia that Contour watches either _or_ but not both. Their default is endpointslices.

Probably best then to not add a test and I don't think it warrants any further API docs - watching endpointslices is now sorta implied via the conformance test.

**Comment by dprotaso** [comment]
tl;dr From Gateway API Meeting Jun 10th

Support FQDN (similar to ExternalName) is too risky and can lead to confused deputy attacks. Hence it's not something the community wants to encourage.

I'm going to close out this issue since there's nothing else actionable here.
